pagefaultgames/pokerogue
# [Bug] Fix first-turn status damage and arena hazards

## What are the changes?

- Adjusted MoveEffectPhase.start() so that ENEMY_SIDE targeted moves no longer occur twice per use in double battles.
- Improved documentation / code readability
- Fixed Toxic / Poison dealing damage immediately when applied.
- Fixed Hazards not persisting through save
- Updated Toxic Orb test

I wasn't able to recreate toxic spikes not being absorbed, or the double message, though I didn't attempt to until after fixing the double battle problem. 

## Why am I doing these changes?
Addresses bugs compiled in #3240 

## What did change?
- Adjusted MoveEffectPhase to only apply the effects of ENEMY_SIDE targeted moves a single time, even in double battles. Previously this caused a bug where you would apply two layers of toxic spikes (and presumably spikes)
- Fixed bug where toxin/poison/burn damage would apply prematurely by adding a new phase scheduled after weather effects to check for and queue status damage.
- As a result of the Poison/Toxic/Burn changes, Toxic and Flame Orb now behave as they should, with the damage not triggering until the second turn. 
- Fixed ArenaTags not being loaded properly after a reset by basing them off of the handling of BattlerTags. Previously the ArenaData class constructor never successfully recovered saved tags from the JSON string. 
- Updated Toxic Orb test to reflect changes.
- ~~Refactored MoveEffectPhase in Phases.ts to improve readability, heavily reducing nested code by breaking the section into various documented helper functions~~ Separated out into #3541 

### Screenshots/Videos
Only one layer of Spikes/Toxic Spikes are applied even in double battles:

https://github.com/user-attachments/assets/ad1dc723-475f-453b-9fa8-8de296159438

Only one status affliction message is displayed per target:

https://github.com/user-attachments/assets/5aabe286-e59a-4393-b570-80ebf1fc4ed5

Hazards now persist through a reload:

https://github.com/user-attachments/assets/97e50058-3215-4ad2-b2cb-690f39fb9173

Toxic Spikes are correctly absorbed by grounded poison types:

https://github.com/user-attachments/assets/2c7845ee-9fd4-4928-8c28-20a11e1d439d

Additionally, Pokemon no longer take damage immediately upon being poisoned by Toxic Spikes. (The damage they take upon entering the field is due to the normal Spikes)

## How to test the changes?

1. Use overrides to force double battles, and to give your starter(s) of choice toxic spikes
2. Use toxic spikes in a wild battle, and then defeat your opponent.
3. Proceed far enough into the next encounter so that the incoming pokemon get poisoned. The text should only display once, and only one set should go down.
4. Refresh the game
5. Load back into the same file, and the spikes should still be present / affect the incoming pokemon
6. Repeat steps 2-5 to test the Badly Poisoned effect
7. If at any point a grounded poison type pokemon enters the field, the spikes should be absorbed.

## Checklist
